draft-irtf-cfrg-vdaf.md

```diff
@@ -3062,6 +3062,129 @@ measurement is sharded. This is provided to the FLP by Prio3.
 | `Field`          | `Field128` ({{fields}}) |
 {: title="Parameters of validity circuit Histogram."}
 
+### Prio3SumVec
+
+This instance of Prio3 supports summing a vector of integers. It has three
+parameters, `length`, `bits`, and `chunk_length`. Each measurement is a vector
+with length equal to the `length` parameter. Each element of the measurement is
+an integer in the range `[0, 2^bits)`. It is RECOMMENDED to set `chunk_length`
+to an integer near to the square root of `length * bits`. The optimal choice for
+any measurement size will vary slightly due to rounding.
+
+This instance uses PrgSha3 ({{prg-sha3}}) as its PRG. Its validity circuit,
+denoted `SumVec`, uses `Field128` ({{fields}}) as its finite field.
+
+Measurements are encoded as a vector of field elements with length
+`length * bits`. The field elements in the encoded vector represent all the bits
+of the measurement vector's elements, consecutively, in LSB to MSB order:
+
+~~~
+def encode(SumVec, measurement: Vec[Unsigned]):
+    if len(measurement) != SumVec.length:
+        raise ERR_INPUT
+
+    encoded = []
+    for elem in measurement:
+        if 0 > elem or elem >= 2 ** SumVec.bits:
+            raise ERR_INPUT
+
+        for l in range(SumVec.bits):
+            encoded.append(Sum.Field((elem >> l) & 1))
+
+    return encoded
+
+def truncate(SumVec, meas):
+    truncated = [SumVec.Field(0) for _ in range(SumVec.length)]
+    for i in range(SumVec.length):
+        for j in range(SumVec.bits):
+            weight = SumVec.Field(1 << j)
+            truncated[i] += weight * meas[i * SumVec.bits + j]
+    return truncated
+
+def decode(SumVec, output, _num_measurements):
+    return [x.as_unsigned() for x in output]
+~~~
+
+This validity circuit uses a `ParallelSum` gadget to achieve a smaller proof
+size. This optimization for "parallel-sum circuits" is described in
+{{BBCGGI19}}, section 4.4. Briefly, for circuits that add up the output of
+multiple identical sub-circuits, it is possible to achieve smaller proof sizes
+(on the order of O(sqrt(n)) instead of O(n)) by packaging more than one such
+subcircuit into a gadget.
+
+The `ParallelSum` gadget is parameterized with an arithmetic subcircuit, and a
+`count` of how many times it evaluates that subcircuit. It takes in a list of
+inputs and passes them through to instances of the subcircuit in the same order.
+It returns the sum of the subcircuit outputs.
+
+~~~
+def ParallelSum(self, inputs):
```

Comment:
* We should assert any necessary preconditions, i.e., `len(inputs)` should be equal to `self.cont * self.subcircuit.ARITY`, correct?
* Would `subgadget` be a more precise term to use? Since it has the an `ARITY` field, we're expecting it to implement the `Gadget` interface.

Reply:
The actual implementation includes a call to `check_gadget_eval()`, which compares the input length to the gadget's arity, which is exactly what we need. This is moreso a symptom of us hiding implementation details in our descriptions of gadgets and validity circuits in the document.

I wanted to avoid calling it a gadget, because it doesn't participate in recording of input wires during evaluation and proof/verification of gadget polynomial consistency. Perhaps we should define a separate `Circuit` interface that both `Valid` and `Gadget` inherit from, that just covers `ARITY`, `DEGREE`, `eval()`, and `eval_poly()`?

Reply:
Agreed on the first point 👍 

To the second point: That's a good point about not overloading the term "gadget". Perhaps all that's needed is a comment clarifying that `subcircuit.ARITY` refers to the arity of the gadget that we're parallelizing. I'm wary of making a bigger change this, as it may end up creating more work when we address https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/226.